libraries/mpxe-store/src/store.c

```diff
@@ -0,0 +1,174 @@
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "critical_section.h"
+#include "gpio.pb-c.h"
+#include "log.h"
+
+#define MAX_STORE_COUNT 64
+#define INVALID_STORE_ID MAX_STORE_COUNT
+
+typedef struct Store {
+  MxStoreType type;
+  void *key;
+  void *store;
+} Store;
+
+// Used to ensure this module is only initialized once
+static bool s_initialized = false;
+
+static Store s_stores[MAX_STORE_COUNT];
+
+// Child to parent fifo - used to talk to harness
+static int s_ctop_fifo;
+
+static StoreFuncs s_func_table[MX_STORE_TYPE__END];
+
+static pthread_mutex_t s_sig_lock;
+
+// signal handler for catching parent
+static void prv_sigusr(int signo) {
+  pthread_mutex_unlock(&s_sig_lock);
+  return;
+}
+
+Store *prv_get_first_empty() {
+  for (uint16_t i = 0; i < MAX_STORE_COUNT; i++) {
+    if (s_stores[i].key == NULL && s_stores[i].store == NULL) {
+      return &s_stores[i];
+    }
+  }
+  return NULL;
+}
+
+static void prv_handle_store_update(uint8_t *buf, int64_t len) {
+  MxStoreUpdate *update = mx_store_update__unpack(NULL, (size_t)len, buf);
+  s_func_table[update->type].update_store(update->msg, update->mask);
+  mx_store_update__free_unpacked(update, NULL);
+}
+
+// handles getting an update from python, runs as thread
+static void *prv_poll_update(void *arg) {
+  // read protos from stdin
+  // compare using second proto as 'mask'
+  // trigger gpio interrupt as necessary
+  struct pollfd pfd = { .fd = STDIN_FILENO, .events = POLLIN };
+  LOG_DEBUG("starting to poll\n");
+  while (true) {
+    int res = poll(&pfd, 1, -1);
+    if (res == -1) {
+      // interrupted
+      continue;
+    } else if (res == 0) {
+      continue;  // nothing to read
+    } else {
+      if (pfd.revents & POLLIN) {
+        static uint8_t buf[4096];  // store may be bigger than this?
+        int64_t len = read(STDIN_FILENO, buf, sizeof(buf));
+        if (len == -1) {
+          LOG_DEBUG("read error while polling\n");
+        }
+        prv_handle_store_update(buf, len);
+      } else {
+        LOG_DEBUG("pollhup\n");
+      }
+    }
+  }
+  return NULL;
+}
+
+void store_config(void) {
+  // do nothing after the first call
+  if (s_initialized) {
+    return;
+  }
+
+  // set up signal handler
+  signal(SIGUSR1, prv_sigusr);
+  pthread_mutex_init(&s_sig_lock, NULL);
+
+  // set up polling thread
+  pthread_t poll_thread;
+  pthread_create(&poll_thread, NULL, prv_poll_update, NULL);
+
+  // set up store pool
+  memset(&s_stores, 0, sizeof(s_stores));
+
+  // open fifo
+  char fifo_path[64];
+  snprintf(fifo_path, sizeof(fifo_path), "/tmp/%d_ctop", getpid());
+  mkfifo(fifo_path, 0666);
+  s_ctop_fifo = open(fifo_path, O_WRONLY);
+
+  s_initialized = true;
+}
+
+void store_register(MxStoreType type, StoreFuncs funcs, void *store, void *key) {
+  s_func_table[type] = funcs;
+  // malloc a proto as a store and return a pointer to it
+  Store *local_store = prv_get_first_empty();
+  if (store == NULL) {
+    return;
+  }
```

Comment:
in theory an outbound-only communication protocol could have a null store since it would just export a message to python but we don't actually do that so yeah I'll just do that